akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala

```diff
@@ -29,60 +28,35 @@ import scala.collection.immutable
-      case c: ChunkedHttp2SubStream => HttpEntity.Chunked(contentType, c.data)
-    }
-
-  def createResponseEntity(contentLength: Long, contentType: ContentType): RequestEntity
-}
-
-private[http2] final case class ByteHttp2SubStream(
+private[http2] case class Http2SubStream(
   initialHeaders:        ParsedHeadersFrame,
-  data:                  Source[ByteString, Any],
+  data:                  Source[Any, Any],
   correlationAttributes: Map[AttributeKey[_], _] = Map.empty
-) extends Http2SubStream {
-  type T = ByteString
-  override def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream =
-    copy(correlationAttributes = attributes)
-
-  def createResponseEntity(contentLength: Long, contentType: ContentType): RequestEntity = {
-    // Ignore trailing headers when content-length is defined
-    if (contentLength == 0) HttpEntity.Empty
-    else if (contentLength > 0) HttpEntity.Default(contentType, contentLength, data)
-    else HttpEntity.Chunked.fromData(contentType, data)
-  }
-}
-
-private[http2] final case class ChunkedHttp2SubStream(
-  initialHeaders:        ParsedHeadersFrame,
-  data:                  Source[HttpEntity.ChunkStreamPart, Any],
-  correlationAttributes: Map[AttributeKey[_], _]                 = Map.empty
-) extends Http2SubStream {
-  type T = HttpEntity.ChunkStreamPart
-
-  override def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream =
-    copy(correlationAttributes = attributes)
-
-  def createResponseEntity(contentLength: Long, contentType: ContentType): RequestEntity =
-    // Ignore trailing headers when content-length is defined
-    if (contentLength == 0) HttpEntity.Empty
-    else if (contentLength > 0) HttpEntity.Default(contentType, contentLength, data.collect { case Chunk(bytes, _) => bytes })
-    else HttpEntity.Chunked(contentType, data)
+) {
+  def streamId: Int = initialHeaders.streamId
+  def hasEntity: Boolean = !initialHeaders.endStream
+
+  def withCorrelationAttributes(newAttributes: Map[AttributeKey[_], _]): Http2SubStream =
+    copy(correlationAttributes = newAttributes)
+
+  def createEntity(contentLength: Long, contentType: ContentType): RequestEntity =
+    if (data == Source.empty || contentLength == 0)
+      if (contentType == ContentTypes.NoContentType) HttpEntity.Empty
+      else HttpEntity.Strict(contentType, ByteString.empty)
+    else if (contentLength > 0) {
+      val byteSource: Source[ByteString, Any] = data.collect {
+        case b: ByteString             => b
+        case HttpEntity.Chunk(data, _) => data
+        // ignore: HttpEntity.LastChunk
+      }
+      HttpEntity.Default(contentType, contentLength, byteSource)
+    } else {
+      val chunkSource: Source[HttpEntity.ChunkStreamPart, Any] = data.map {
+        case b: ByteString                 => HttpEntity.Chunk(b)
+        case p: HttpEntity.ChunkStreamPart => p
+      }
+      HttpEntity.Chunked(contentType, chunkSource)
+    }
```

Comment:
Aah, so the streams can just mix and match raw bytestrings and chunks, postponing selecting the right wrapping until here. That provides a lot of flexibility. Would we expect an impact on performance?

Reply:
WDYM, because of the matching?

Reply:
adding the additional `map` to the stream to do the wrapping here - though I guess that's all fused so it shouldn't matter?

Reply:
It could matter but at least it's run in a user stream, i.e. not starving the HTTP/2 processing itself.

I'm a bit more concerned about the general performance implications of putting each frame separately through the whole HTTP/2 pipeline on input and output side. It seems that when TLS is used, it doesn't really matter because even on JDK 11, HTTP/2 seems to be TLS/crypto constrained but that may change in the future and we should continue to monitor potential performance issues.